src/backend/logger/logger.ts

```diff
@@ -361,43 +362,76 @@ const logsWriters: Record<string, LogWriter> = {}
 
 // Base abstract class for all LogWriters
 class LogWriter {
-  queue: string[]
+  queue: (string | Promise<string | string[]>)[]
   initialized: boolean
   filePath: string
   timeoutId: NodeJS.Timeout | undefined
+  processing: boolean
 
   constructor() {
     this.initialized = false
     this.queue = []
     this.filePath = ''
+    this.processing = false
 
     if (new.target === LogWriter) {
       throw new Error('LogWriter is an abstract class')
     }
   }
 
-  logMessage(message: string) {
+  /**
+   * Append a message to the queue
+   * @param message string or promise that returns a string or string[]
+   */
+  logMessage(message: string | Promise<string | string[]>) {
     // push messages to append to the log
     this.queue.push(message)
 
-    // if the logger is initialized and we don't have a timeout,
-    // append the message and start a timeout
+    // if the logger is initialized and we don't have a timeout
+    // and we are not proccesing the previous batch, write a new batch
     //
     // otherwise it means there's a timeout already running that will
-    // write the elements in the queue in a second
-    if (this.initialized && !this.timeoutId) this.appendMessages()
+    // write the elements in the queue in a second or that we are processing
+    // promises
+    if (this.initialized && !this.processing && !this.timeoutId)
+      this.appendMessages()
   }
 
   async appendMessages() {
-    const messagesToWrite = this.queue
+    const itemsInQueue = this.queue
 
     // clear pending message if any
     this.queue = []
 
     // clear timeout if any
     delete this.timeoutId
 
-    if (!messagesToWrite?.length) return
+    if (!itemsInQueue?.length) return
+
+    this.processing = true
+
+    // process items in queue, if they are promises we wait
+    // for them so we can write them in the right order
+    let messagesToWrite: string[] = []
+    for (const item of itemsInQueue) {
+      if (typeof item === 'string') {
+        messagesToWrite.push(item)
+      } else {
+        try {
+          let result = await item
+
+          // support a promise returning a string or an array of strings
+          if (typeof result === 'string') {
+            result = [result]
+          }
+          messagesToWrite = messagesToWrite.concat(result)
+        } catch (error) {
+          logError(error, LogPrefix.Backend)
+        }
+      }
```

Comment:
Assuming we keep this code (see my other comment), this can still be simplified:
```suggestion
      try {
        const result = await item
        messagesToWrite = Array.isArray(result) ? result : [result]
      } catch (error) {
        logError(error, LogPrefix.Backend)
      }
```

Reply:
I pushed this refactor

I do have to keep the contact part cause I'm processing multiple elements of the queue (and each element can be either a string or an array of strings)

Reply:
Oh right, of course